pkcs11: remove existing PKCS#11 keys to avoid label conflicts

detsch opened the pull request:
The current implementation allows multiple key pairs to be stored in the HSM using the same "tls" label, which leads to certificate errors inside aklite. Even a previous incomplete lmp-device-register execution might leave these keys behind.

This commit makes sure there are no other stored keys using the same label before creating a new key pair during the registration process.

mike-sul commented on src/pkcs11.cpp:
```diff
@@ -204,6 +235,9 @@ int pkcs11_create_csr(const lmp_options &opt, string &pkey, string &csr)
 	if (PKCS11_login(slot, 0, opt.hsm_pin.c_str()))
 		leave;
 
+	if (remove_keys(slot->token, hsm_cfg.tls_lbl))
```
If an initial registration was successful, and then for some reason a user decides to re-register a device again, and the re-registration fails then a device ends up without valid "keys".
I am not sure that this is critical, because a device ultimately gets valid "keys" once it is successfully re-registered.

detsch replied:
The user would have had to remove `/var/sota/` content in order to be able to try to re-register, so I believe removing the keys is fine, even if something goes wrong later on.
I should probably remove existing certificates with "client" label as well, since they would be referring to key pairs that are now removed, and won't be usable anymore. I'll add that once I'm back later today.
@ldts any comments on that and on the general fix? I can't add reviewers to the PR.

mike-sul replied:
>The user would have had to remove /var/sota/ content in order to be able to try to re-register, so I believe removing the keys is fine, even if something goes wrong later on.

Not exactly, this is enough to remove just `sql.db` to re-register a device. 
Having said that I don't see any use-case when removing the key can cause some trouble providing that we document this detail ("device re-registration is not an atomic transaction, its failure doesn't mean that a device is the same state as before the re-registration").

Maybe @doanac can think of an use-case when it can be critical.

doanac replied:
I think from a validation standpoint, we have to protect for this:

 * lmp-device-register runs successfully. aklite daemon is up
 * time goes by
 * aklite is stopped for some reason. At this point people tend to re-run lmp-device-register w/o really thinking through the consequences. As long as lmp-device-register exits here cleanly and doesn't remove keys we are fine.
 * user does an `rm /var/sota/sql.db` cause they really want to re-register
 * lmp-device-register should be able to work and generate new keys

mike-sul replied:
>aklite is stopped for some reason. At this point people tend to re-run lmp-device-register w/o really thinking through the consequences. As long as lmp-device-register exits here cleanly and doesn't remove keys we are fine.

It is so, as long as `sql.db` is not removed.

mike-sul replied:
>user does an rm /var/sota/sql.db cause they really want to re-register
> lmp-device-register should be able to work and generate new keys

What if it doesn't work (re-registration fail), is it fine that the key is removed?
I think, it is OK. Since a user removes `sql.db` he/she is in "advanced" mode and should continue with device re-registration attempts until it succeeds.

ricardosalveti replied:
Yeah, at this stage the user is already doing a more advanced flow. Without sql.db we can't really know what to do.

I know there could be use cases such as when the user performs a factory reset but decides to keep the hsm specific keys, but for that we would need to restore sota.toml and currently sql.db as well in order for the device to come up in a managed stage again after factory reset.

Saying that, I think the current logic here is fine.

mike-sul commented:
LGTM. Since @ldts owns lmp-dev-reg now, I defer approval to him.

detsch commented:
I've now added code for removing existing certificates as well, in order to better support a re-registration scenario (not just a re-try after a failed attempt).

mike-sul commented:
> I've now added code for removing existing certificates as well, in order to better support a re-registration scenario too (not just a re-try after a failed attempt).

I wonder if it's possible to unlabel/(label differently) or remove or move or backup the whole "container" (token) instead of removing each item separately.

I am not familiar enough with the current lmp-dev-reg source code to approve or disapprove the PR, I defer it to @ldts.

ricardosalveti commented:
Found it a bit weird that we could have multiple keys with the same ids/labels, but that is indeed allowed by the spec (https://github.com/OpenSC/OpenSC/issues/1613):

PKCS#11-2.40 says:

"CKA_LABEL RFC2279 string Description of the object (default empty)." and "The CKA_LABEL attribute is intended to assist users in browsing." It does not say anything about uniqueness.

In Section "4.6.3 X.509 public key certificate objects" it says:
"CKA_ID Byte array Key identifier for public/private key pair (default empty)." and "The CKA_ID attribute is intended as a means of distinguishing multiple public-key/private-key pairs held
by the same subject (whether stored in the same token or not). (Since the keys are distinguished by
subject name as well as identifier, it is possible that keys for different subjects may have the same
CKA_ID value without introducing any ambiguity.)"

Then in section "4.7.2 Overview" it says: "The CKA_ID field is intended to distinguish among multiple keys. In the case of public and private keys, this field assists in handling multiple keys held by the same subject; the key identifier for a public key and its corresponding private key should be the same. The key identifier should also be the same as for the corresponding certificate, if one exists. Cryptoki does not enforce these associations, however. (See Section 4.6 for further commentary.)"

ricardosalveti commented:
So in order to avoid confusion it is indeed better to remove them when the token is already initialized (since we don't re-init it anymore as that would erase all keys/objects).

ricardosalveti commented:
> I wonder if it's possible to unlabel/(label differently) or remove or move or backup the whole "container" (token) instead of removing each item separately.
> 
> I am not familiar enough with the current lmp-dev-reg source code to approve or disapprove the PR, I defer it to @ldts.

At this stage I don't think there would be value in keeping older keys/certs, as the available space for keys/certs is also limited in RPMB/HSM.

Since this is a token used by aktualizr, and we're using keys/certs that are also controlled/used by it, we should be fine to remove to avoid duplicates.

ricardosalveti reviewed (approved):
LGTM

ldts commented:
Instead of acting on it, the error should  detected and the  process aborted (just for simplicity and consistency with current behavior). Otherwise we should also be doing some error handling on all the openssl and pkcs11 functions and it becomes a pain. We should keep this binary simple.

So the way I would proceed is:
1) detect this as an error and abort
2) have one --force option to remove the content of /var/sota _and_ clear the unexpected in pkcs#11 storage or anywhere else.

detsch commented:
Updated the commit based on @ldts suggestions. @rsalveti, @mike-sul please take a look again, since the code has changed significantly.

ldts commented on src/pkcs11.cpp:
```diff
@@ -139,6 +139,87 @@ static int pkcs11_initialize(const lmp_options &opt)
+ * Verify if the HSM has certificates with the given label.
+ * Abort execution or remove them depending on the remove option
+ */
+static int locate_certs(PKCS11_TOKEN *token, string label, bool remove)
```
instead of locate_certs with an option to remove could we do it in two steps
something like

```
if (certificate_exists(...))
   if (force)
        delete_certificate()
   else
        return -1
```

detsch replied:
In that case, if `force=true`, we would have to iterate over the certificates twice, since there might be more that 1 certificate with the conflicting label. Should I change it anyway?

ldts replied:
yes please. I am not too concerned about performance during registration (a few miliseconds more when we are going to the cloud doesnt really matter IMO).

Also could you do the changes in **pkcs11_check_hsm()** instead

ldts replied:
also notice that if the token is not found there is no point on checking for the presence of keys or certificates

detsch replied:
> instead of locate_certs with an option to remove could we do it in two steps

> Also could you do the changes in pkcs11_check_hsm() instead

> also notice that if the token is not found there is no point on checking for the presence of keys or certificates

Handled the 3 suggestions above. (Missed the `pkcs11_check_hsm` one initially, so I had to made a second force push.)

mike-sul commented:
> Updated the commit based on @ldts suggestions. @rsalveti, @mike-sul please take a look again, since the code has changed significantly.

LGTM.

I wonder if it would make sense to remove a private key if the backend returns error since we already have the function to locate and remove it from HSM?

I can image the use-case when a device hasn't been registered yet and the first attempt fails. Then, the second attempt will require specifying of the  `--force` option which looks a bit odd since a device never been registered from a user perspective.

detsch commented:
> I wonder if it would make sense to remove a private key if the backend returns error since we already have the function to locate and remove it from HSM?
> 
> I can image the use-case when a device hasn't been registered yet and the first attempt fails. Then, the second attempt will require specifying of the `--force` option which looks a bit odd since a device never been registered from a user perspective.

I agree. Cleaning up the key pair if the backend returns an error is better than requiring the `--force` in the next attempt. I'll add that logic.

ldts commented:
if we are going to start doing proper cleanups then I suggest we:
1. make sure the program execution **never leaves anything behind on error**, whether this is in the HSM or in the filesystem.
1.1  this will need a signal handler to be sure in case the execution is aborted.
2. and coming back to my point, make sure any new execution never deletes anything that was not created by it **unless explicitly asked.**

detsch commented:
I've created a new commit for cleaning up the created pkcs11 keys  if an error occurs. 

If we agree on improvements on the cleanup process, as suggested by @ldts , (cleanup filesystem +  add signal handlers), I believe we can to it on top of the current changes.

Current code is not supposed to remove anything not created by the current execution, unless `--force 1` is used.

mike-sul commented:
> If we agree on improvements on the cleanup process, as suggested by @ldts , (cleanup filesystem + add signal handlers), I believe we can to it on top of the current changes.

I think it can be even in a separate/new PR.
I am not sure that there is anything to clean in the HSM case. If a root.crt write is atomic and occurs only after successful registration then it should not appear on a file system if some error happens. 
If a response from backend is successful and root.crt and client.crt are not written to or removed from a disk if some interrupting signal happens then device is considered as registered at the backend. So, the following registration retry fails because of name/uuid conflict. So, a user or user software around lmp-dev-reg has to deal with the given negative scenario and delete the device from backend prior to the next registration attempt.
This my argument against cleaning anything on fs or handling signals.

ldts commented:
I dont see why it matters what the back-end thinks: the device should either have everything it needs to function (a fully populated sota and the necessary keys) or nothing at all. 

For the use case that you describe, the abort signal handler should send a de-register to the backend in the case that populate_sote_dir fails. Or at least log the error persistently somewhere that can be checked before the next run.

ldts reviewed (changes requested):
Not cleaning up up keys was an understood design choice since similar keys could have existed anyways (allowed by the spec)

What is being done in this commit has zero guarantees that the created keys will be deleted in case of errors since it is only checking registration errors.

Please cover all error cases from the moment keys are created until the process exiting (including the scenario where a kill signal is sent to the process). Then we know that created keys should always be deleted.

At the very least - since we are doing house keeping- we now should have a single exit point to keep it all simple: add an **error** label at the bottom of `main` to do the house keeping from `create_csr`, `auth_register_device` and `populate_sota_dir`.  You could also have an enum with different types of errors or actions for house_keeping to perform.  then just remove the calls to exit (there are a couple of curl.hpp and auth.cpp that need to be removed)

ldts commented on src/pkcs11.cpp:
```diff
@@ -139,6 +139,100 @@ static int pkcs11_initialize(const lmp_options &opt)
+	unsigned int n = 0;
+
+	if (PKCS11_enumerate_certs(token, &cert, &n) || !n)
+		return false;
```
better to return 0

ldts commented on src/pkcs11.cpp:
```diff
@@ -139,6 +139,100 @@ static int pkcs11_initialize(const lmp_options &opt)
+		if (cert->label == label)
+			return true;
+
+	return 0;
```
return false

ldts commented:
for cleanups also maybe in pkcs11.cpp  have a new leave_cleanup macro that gotos to a cleanup label at the bottom of the current function...then if the function is creating persistent stuff we know that on errors it wont be leaving stuff behind.

detsch commented:
Added a new commit on top of the previous two. A few comments:
- The main function and the signal handling function are responsible for cleaning created objects
- Additionally, a function that create objects and an error is detected after the object was created, still inside the same function, a cleanup of the specific objects is done before going back to main. This is just an a redundant mechanism, since, the main function will still do the general cleanup after the function returns
- Additional scopes had to be added to the token requested in `get_oauth_token`. Was just `devices:create`, now we ask for `devices:delete` and `devices:read-update` as well, allowing the device to be unregistered
- I've added a few static variables intended to be shared only between main and signal handling function

mike-sul commented:
> * Additional scopes had to be added to the token requested in `get_oauth_token`. Was just `devices:create`, now we ask for `devices:delete` and `devices:read-update` as well, allowing the device to be unregistered

I suggest to get device deletion out of this effort scope as it triggers lots of questions/concerns. It will help to finalize and merge this PR as it's highly useful even without the device deletion functionality.

detsch commented:
> I suggest to get device deletion out of this effort scope as it triggers lots of questions/concerns. It will help to finalize and merge this PR as it's highly useful even without the device deletion functionality.

@ldts, what do you think? Would it be acceptable to skip device de-registration for now?

ricardosalveti commented:
> > I suggest to get device deletion out of this effort scope as it triggers lots of questions/concerns. It will help to finalize and merge this PR as it's highly useful even without the device deletion functionality.
> 
> @ldts, what do you think? Would it be acceptable to skip device de-registration for now?

@detsch lets go ahead and skip the de-registration for now until we can align on a proper implementation with the backend folks.

detsch commented:
Dropped the de-registration logic. Cleanup is now restricted to local data (files and HSM objects), and happens on every error or SIGSEGV / SIGINT signals.

vkhoroz reviewed (approved):
LGTM.

But this is definitely something @ldts needs to approve.

ricardosalveti reviewed (approved):
LGTM

mike-sul commented on src/main.cpp:
```diff
@@ -47,6 +48,40 @@ static bool ends_with(const std::string &s, const std::string &suffix)
+	if (!opt.force) {
+		cerr << "ERROR: Device already registered in "
+		     << opt.sota_dir << endl;
+		cerr << "Re-run with --force 1 to remove existing registration "
```
Maybe we should advise a user to delete a device if he/she would like to re-register the device with the same name/ID (something like run `fioctl device delete`).

detsch replied:
At that point we are not sure about the situation at the backend. And we would have to tell about delete-denied as well. Perhaps a better location for that hint is a HTTP CONFLICT is returned on a registration attempt. But I would rather do improvements in that area as a separate effort, to avoid dragging the PR longer.

mike-sul commented on src/auth.cpp:
```diff
@@ -10,6 +10,17 @@
+	if (resp.data().length())
+		cerr << resp.data() << endl;
+
+	for (auto it: resp)
```
Just really nitpicking :). This loop would be slightly better if defined as `for (const auto& it: resp)`.

detsch replied:
Updated the code accordingly

mike-sul commented on src/main.cpp:
```diff
@@ -299,32 +325,71 @@ int create_csr(const lmp_options &opt, string &key, string &csr)
+
+static void cleanup_generated_data()
+{
+	cout << "Cleaning up partial registration before leaving" << endl;
```
A signal may arrive after a device has successfully registered and before all keys/certs are written to a disk.
So, if are removing partially stored stuff for the registered device maybe it makes sense to print some message so a user is aware that a device was actually registered at the backend?

detsch replied:
We no longer have flags telling at which point the execution was stopped. The logs give some clue, and the user can infer from there if the registration at the backend has succeed. Also, errors are not likely to happen after the registration is successful. Like for https://github.com/foundriesio/lmp-device-register/pull/58#discussion_r1338309645 improvements in the messaging can be part of a separate effort, if we feel is required.

mike-sul commented:
LGTM, I just have minor suggestions/notes. I defer to @ldts a detailed review.

ldts commented:
review is going on via patches

ldts commented on src/pkcs11.cpp:
```diff
@@ -502,7 +524,8 @@ int pkcs11_remove_keys(lmp_options &opt)
 		leave;
 
-	remove_keys(slot->token, hsm_cfg.tls_lbl);
+	remove_lmp_keys(slot->token);
```
we can check the return value and leave on error

detsch replied:
I was about to  comment on that, since I intentionally reverted  your change. Since that happens inside the cleanup routine, I believe is better to just keep removing what we can, and not skip PKCS11_logout if an error happens here. I can keep your changes if you do not agree with that.

ldts replied:
I think it is better to be consistent on how we handle errors than doing our best and then get random results to analyze. This way we know exactly where we are failing and why...so I'd rather fail early really.

I sent you another patch with some more stuff.

detsch replied:
Inside `populate_sota_dir` we either need a new macro instead of `leave_exit`, or we need to keep `int ret` defined. I'm re-adding `int ret` for now to proceed with testing.

detsch replied:
Based on your other changes I'm replacing `leave_exit` with `goto error` in `populate_sota_dir` to fix the build.

detsch replied:
Applied your changes + the required fixes to allow code to compile.
